lib/p2p/Parser.ts

```diff
@@ -60,20 +60,30 @@ interface Parser {
 
 /** Protocol packet parser */
 class Parser extends EventEmitter {
+  private pending = '';
+
+  constructor(private delimiter: string) {
+    super();
+  }
+
   public feed = (data: Buffer): void => {
-    const dataStr = data.toString();
-    const packetLines = dataStr.split('\r\n');
-    packetLines.forEach((packetStr) => {
-      if (!packetStr) {
-        return;
-      }
-      try {
-        const packet = fromRaw(packetStr);
-        this.emit('packet', packet);
-      } catch (err) {
-        this.emit('error', err);
-      }
-    });
+    this.pending += data.toString();
+    const index = this.pending.indexOf(this.delimiter);
```

Comment:
I have a minor performance concern here where `indexOf` will get slower and slower as `pending` gets larger, but I realize it's not as simple as checking the `data.toString()` for the delimiter because the delimiter could be split between two chunks. I'm pretty sure this is O(n^2) complexity with the size of the packet and something we'll be doing very often so I'm just concerned and maybe we should at least make a note to revisit this.

One solution that just came to mind is tracking the last received chunk and the current chunk, and concatenating those then checking the result for the delimiter. We'd clear the last chunk on creating a packet.

Not necessary for this PR, but something to think about.

Reply:
Your solution could improve it, you can open a separate issue for that if you want. Not high priority at the moment.